src/darwin/Framework/CHIP/MTRDeviceController.mm

```diff
@@ -654,6 +656,68 @@ - (void)setPairingDelegate:(id<MTRDevicePairingDelegate>)delegate queue:(dispatc
     });
 }
 
+- (void)setNocChainIssuer:(id<MTRNOCChainIssuer>)nocChainIssuer
+{
+    VerifyOrReturn([self checkIsRunning]);
+
+    dispatch_sync(_chipWorkQueue, ^{
+        VerifyOrReturn([self checkIsRunning]);
+
+        self->_operationalCredentialsDelegate->SetNocChainIssuer(nocChainIssuer);
+    });
+}
+
+- (NSNumber *)onNOCChainGenerationComplete:(NSData *)operationalCertificate
+                   intermediateCertificate:(NSData *)intermediateCertificate
+                           rootCertificate:(NSData *)rootCertificate
+                                       ipk:(NSData *)ipk
+                              adminSubject:(NSNumber *)adminSubject
+                                     error:(NSError * __autoreleasing *)error
+{
+    VerifyOrReturnValue([self checkIsRunning:error], [NSNumber numberWithUnsignedInt:CHIP_ERROR_INCORRECT_STATE.AsInteger()]);
+    VerifyOrReturnValue(operationalCertificate != nil, [NSNumber numberWithUnsignedInt:CHIP_ERROR_INVALID_ARGUMENT.AsInteger()]);
+    VerifyOrReturnValue(intermediateCertificate != nil, [NSNumber numberWithUnsignedInt:CHIP_ERROR_INVALID_ARGUMENT.AsInteger()]);
+    VerifyOrReturnValue(rootCertificate != nil, [NSNumber numberWithUnsignedInt:CHIP_ERROR_INVALID_ARGUMENT.AsInteger()]);
+
+    // use ipk and adminSubject from CommissioningParameters if not passed in
+    chip::Optional<chip::Controller::CommissioningParameters> commissioningParameters
+        = _cppCommissioner->GetCommissioningParameters();
+    VerifyOrReturnValue(
+        commissioningParameters.HasValue(), [NSNumber numberWithUnsignedInt:CHIP_ERROR_INCORRECT_STATE.AsInteger()]);
+
+    chip::Optional<chip::Crypto::AesCcm128KeySpan> ipkOptional;
+    uint8_t ipkValue[chip::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES];
+    chip::Crypto::AesCcm128KeySpan ipkTempSpan(ipkValue);
+    if (ipk != nil) {
+        VerifyOrReturnValue(
+            [ipk length] == sizeof(ipkValue), [NSNumber numberWithUnsignedInt:CHIP_ERROR_INCORRECT_STATE.AsInteger()]);
+        memcpy(&ipkValue[0], [ipk bytes], [ipk length]);
+        ipkOptional.SetValue(ipkTempSpan);
+    } else if (commissioningParameters.Value().GetIpk().HasValue()) {
+        ipkOptional.SetValue(commissioningParameters.Value().GetIpk().Value());
+    }
+
+    chip::Optional<chip::NodeId> adminSubjectOptional;
+    if (adminSubject == nil) {
+        // if no value passed in, use value from CommissioningParameters
+        adminSubjectOptional.SetValue(commissioningParameters.Value().GetAdminSubject().ValueOr(chip::kUndefinedNodeId));
```

Comment:
`kUndefinedNodeId` is not OK here, I'd think....  Why would we not leave this as no-value if we have no value?  So:
```
adminSubjectOptional = commissioningParameters.Value().GetAdminSubject();
```

Reply:
Updated.